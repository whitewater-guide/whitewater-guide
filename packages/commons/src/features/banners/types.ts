import { Overwrite } from 'type-zoo';
import { NamedNode, Node } from '../../core';
import { Group } from '../groups';
import { Region } from '../regions';
import { Connection } from '../types';

export enum BannerPlacement {
  MOBILE_SECTION_DESCRIPTION = 'mobileSectionDescription',
  MOBILE_SECTION_ROW = 'mobileSectionRow',
  MOBILE_SECTION_MEDIA = 'mobileSectionMedia',
  MOBILE_REGION_DESCRIPTION = 'mobileRegionDescription',
}

export enum BannerKind {
  Image = 'Image',
  WebView = 'WebView',
}

export interface BannerSource {
  kind: BannerKind;
  ratio: number | null;
  src: string | null;
}

export interface Banner extends NamedNode {
  enabled: boolean;
  slug: string;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
  // --- connections
  regions?: Connection<Region>;
  groups?: Connection<Group>;
  // ---- transient
  deleted?: boolean;
}

export interface BannerInput {
  id: string | null;
  slug: string;
  name: string;
  enabled: boolean;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: {[key: string]: any} | null;
  // --- connections
  regions: Node[];
  groups: Node[];
}

export type BannerFormInput = Overwrite<BannerInput, { extras: string | null }>;

export function isBanner(node: Node): node is Banner {
  return node.__typename === 'Banner';
}
