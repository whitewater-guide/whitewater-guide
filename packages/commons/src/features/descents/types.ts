import { Timestamped, PageInfo } from '../../apollo';
import { Section } from '../sections';
import { User } from '../users';

export interface DescentLevel {
  __typename?: 'Level';
  unit?: string | null;
  value?: number | null;
}

export interface DescentLevelInput {
  unit?: string | null;
  value: number;
}

export interface Descent extends Timestamped {
  __typename?: 'Descent';
  id: string;
  user: User;
  section: Section;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevel | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentEdge<TCursor = any> {
  __typename?: 'DescentEdge';
  node: Descent;
  cursor: TCursor;
}

export interface DescentInput {
  id?: string | null;

  sectionId: string;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevelInput | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentsConnection<TCursor = any> {
  __typename?: 'DescentsConnection';
  edges: DescentEdge[];
  pageInfo: PageInfo<TCursor>;
}

export interface DescentsFilter {
  startDate?: string | null;
  endDate?: string | null;
  difficulty?: [number, number] | null;
  userId?: string | null;
  sectionId?: string | null;
  sectionName?: string | null;
}
