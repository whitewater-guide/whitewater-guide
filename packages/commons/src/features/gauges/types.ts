import { NamedNode, Node, Timestamped } from '../../core';
import { HarvestStatus, Measurement } from '../measurements';
import { Point, PointInput } from '../points';
import { Source } from '../sources';

export interface Gauge extends NamedNode, Timestamped {
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
  enabled: boolean;
  location: Point;
  source: Source;
  // measurements?: Measurement[];
  lastMeasurement: Measurement | null;
  status: HarvestStatus | null;
}

export interface GaugeInput {
  id: string | null;
  name: string;
  source: Node;
  location: PointInput | null;
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
}
