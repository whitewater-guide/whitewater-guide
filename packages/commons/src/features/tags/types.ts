import { NamedNode } from '../../apollo';

export enum TagCategory {
  kayaking = 'kayaking',
  hazards = 'hazards',
  supply = 'supply',
  misc = 'misc',
}

export enum TagSelection {
  SELECTED = 'selected',
  DESELECTED = 'deselected',
  NONE = 'none',
}

export const TagSelections = [
  TagSelection.NONE,
  TagSelection.SELECTED,
  TagSelection.DESELECTED,
];

export const TAG_CATEGORIES = ['kayaking', 'hazards', 'supply', 'misc'];

export interface Tag extends NamedNode {
  category: TagCategory;
}

export interface SelectableTag extends Tag {
  selection?: TagSelection;
}

export interface TagInput {
  id: string;
  name: string;
  category: TagCategory;
}

export interface WithTags {
  tags: Tag[];
  tagsLoading: boolean;
}
