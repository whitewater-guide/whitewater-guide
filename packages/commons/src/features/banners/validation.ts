import * as yup from 'yup';
import { yupTypes } from '../../validation';
import { BannerKind, BannerPlacement } from './types';

const BannerSourceSchema = yup
  .object({
    kind: yup.mixed().oneOf(Object.values(BannerKind)),
    ratio: yup
      .number()
      .moreThan(3)
      .lessThan(10)
      .nullable(true),
    src: yupTypes.nonEmptyString(),
  })
  .noUnknown();

export const BannerInputSchema = yup
  .object({
    id: yupTypes.uuid().nullable(),
    slug: yupTypes.slug(),
    name: yupTypes.nonEmptyString(),
    priority: yup.number().integer(),
    enabled: yup.bool(),
    placement: yup.mixed().oneOf(Object.values(BannerPlacement)),
    source: BannerSourceSchema.clone(),
    link: yupTypes.https().nullable(true),
    extras: yup.object().nullable(true),
    regions: yup.array(yupTypes.node()),
    groups: yup.array(yupTypes.node()),
  })
  .strict(true)
  .noUnknown(true);
