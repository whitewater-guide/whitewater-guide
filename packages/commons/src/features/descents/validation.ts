import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';
import { DescentLevelInput, DescentInput } from './types';

export const DescentLevelInputSchema = yup
  .object<DescentLevelInput>({
    unit: yup
      .string()
      .max(20)
      .notRequired()
      .nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema = yup
  .object<DescentInput>({
    id: yupTypes
      .uuid()
      .notRequired()
      .nullable(),
    sectionId: yupTypes
      .uuid()
      .required()
      .nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup
      .number()
      .integer()
      .notRequired()
      .nullable(),
    level: DescentLevelInputSchema.clone()
      .notRequired()
      .nullable(),
    comment: yup
      .string()
      .notRequired()
      .nullable(),
    public: yup
      .bool()
      .notRequired()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
