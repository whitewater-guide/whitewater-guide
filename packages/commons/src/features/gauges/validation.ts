import { baseStruct } from '../../utils/validation';
import { PointInputStruct } from '../points';

const GaugeInputFields = {
  id: baseStruct.union(['uuid', 'null']),
  name: 'nonEmptyString',
  code: 'nonEmptyVarchar',
  levelUnit: 'nonEmptyVarchar|null',
  flowUnit: 'nonEmptyVarchar|null',
  location: baseStruct.union([PointInputStruct, 'null']),
  requestParams: 'object|null',
  cron: baseStruct.union(['cron', 'null']),
  url: baseStruct.union(['url', 'null', baseStruct.literal('')]),
  source: 'node',
};

export const GaugeInputStruct = baseStruct.object(GaugeInputFields);

export const GaugeFormStruct = baseStruct.object({
  ...GaugeInputFields,
  requestParams: baseStruct.union(['jsonString', 'null']),
});
