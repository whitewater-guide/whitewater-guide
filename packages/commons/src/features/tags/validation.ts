import * as yup from 'yup';
import { yupTypes } from '../../validation';
import { TAG_CATEGORIES, TagInput } from './types';

export const TagInputSchema = yup
  .object<TagInput>({
    id: yupTypes.slug(),
    name: yupTypes.nonEmptyString(),
    category: yup.mixed().oneOf(TAG_CATEGORIES),
  })
  .strict(true)
  .noUnknown();
