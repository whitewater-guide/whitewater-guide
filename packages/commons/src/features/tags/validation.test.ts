import { createSafeValidator } from '../../validation';
import { TagCategory, TagInput } from './types';
import { TagInputSchema } from './validation';

const validator = createSafeValidator(TagInputSchema);

type TestValue = [string, TagInput];

const correct: TagInput = {
  id: 'sometag',
  name: 'Some tag',
  category: TagCategory.hazards,
};

const correctValues: TestValue[] = [
  ['full value', correct],
  ['legacy id 1', { ...correct, id: 'Must-run' }],
  ['legacy id 2', { ...correct, id: 'man_made' }],
  ['legacy id 3', { ...correct, id: '4x4' }],
];

const incorrectValues: TestValue[] = [
  ['empty name', { ...correct, name: '' }],
  ['bad category', { ...correct, category: 'foo' as any }],
  ['bad id 1', { ...correct, id: 'aa' }],
  ['bad id 2', { ...correct, id: 'яяяяяя' }],
  ['bad id 3', { ...correct, id: 'foo bar' }],
  ['extra fields', { ...correct, foo: 'foo bar' } as any],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  const error = validator(value);
  expect(error).not.toBeNull();
  expect(error).toMatchSnapshot();
});
