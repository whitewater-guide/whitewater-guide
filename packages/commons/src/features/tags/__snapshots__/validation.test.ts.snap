// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for bad category 1`] = `
Object {
  "category": Object {
    "key": "yup:mixed.oneOf",
    "options": Object {
      "values": "kayaking, hazards, supply, misc",
    },
  },
}
`;

exports[`should be invalid for bad id 1 1`] = `
Object {
  "id": "yup:string.slug",
}
`;

exports[`should be invalid for bad id 2 1`] = `
Object {
  "id": "yup:string.slug",
}
`;

exports[`should be invalid for bad id 3 1`] = `
Object {
  "id": "yup:string.slug",
}
`;

exports[`should be invalid for empty name 1`] = `
Object {
  "name": "yup:string.nonEmpty",
}
`;

exports[`should be invalid for extra fields 1`] = `
Object {
  "undefined": Object {
    "key": "yup:object.noUnknown",
    "options": Object {
      "path": "this",
    },
  },
}
`;
